File: claude-code-sdk/src/main/kotlin/com/claudecodeplus/sdk/mcp/McpServer.kt
package com.claudecodeplus.sdk.mcp

import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonPrimitive

/**
 * MCP Server接口 - 所有MCP服务器的基础接口
 * 
 * 提供了创建自定义MCP工具服务器的标准接口，支持工具列表查询和工具调用执行。
 */
interface McpServer {
    /**
     * 服务器名称，用作标识符
     */
    val name: String
    
    /**
     * 服务器版本
     */
    val version: String
    
    /**
     * 服务器描述
     */
    val description: String
    
    /**
     * 列出所有可用工具
     */
    suspend fun listTools(): List<ToolDefinition>
    
    /**
     * 调用指定的工具
     */
    suspend fun callTool(toolName: String, arguments: Map<String, Any>): ToolResult
}

/**
 * 工具定义
 */
data class ToolDefinition(
    val name: String,
    val description: String,
    val inputSchema: Map<String, Any> = emptyMap()
) {
    companion object {
        /**
         * 创建简单的工具定义
         */
        fun simple(name: String, description: String) = ToolDefinition(
            name = name,
            description = description,
            inputSchema = mapOf(
                "type" to "object",
                "properties" to emptyMap<String, Any>()
            )
        )
        
        /**
         * 从参数类型创建工具定义（兼容旧版本）
         */
        fun withParameters(
            name: String, 
            description: String, 
            parameters: Map<String, ParameterType>
        ) = withParameterInfo(
            name = name,
            description = description,
            parameters = parameters.mapValues { (_, type) ->
                ParameterInfo(type = type)
            }
        )
        
        /**
         * 从参数信息创建工具定义（支持描述）
         */
        fun withParameterInfo(
            name: String, 
            description: String, 
            parameters: Map<String, ParameterInfo>
        ) = ToolDefinition(
            name = name,
            description = description,
            inputSchema = mapOf(
                "type" to "object",
                "properties" to parameters.mapValues { (_, paramInfo) ->
                    val schema = mutableMapOf<String, Any>()
                    
                    // 设置基础类型
                    when (paramInfo.type) {
                        ParameterType.STRING -> schema["type"] = "string"
                        ParameterType.NUMBER -> schema["type"] = "number" 
                        ParameterType.BOOLEAN -> schema["type"] = "boolean"
                        ParameterType.ARRAY -> schema["type"] = "array"
                        ParameterType.OBJECT -> schema["type"] = "object"
                    }
                    
                    // 添加描述信息
                    if (paramInfo.description.isNotEmpty()) {
                        schema["description"] = paramInfo.description
                    }
                    
                    schema.toMap()
                },
                "required" to parameters.keys.toList() // 所有参数都是必需的，除非有默认值
            )
        )
    }
}

/**
 * 参数类型枚举
 */
enum class ParameterType {
    STRING, NUMBER, BOOLEAN, ARRAY, OBJECT
}

/**
 * 参数信息类 - 只包含类型和描述信息
 */
data class ParameterInfo(
    val type: ParameterType,
    val description: String = ""
)

/**
 * 工具执行结果
 */
sealed class ToolResult {
    abstract val content: List<ContentItem>
    abstract val isError: Boolean
    
    /**
     * 成功结果
     */
    data class Success(
        override val content: List<ContentItem>,
        val metadata: Map<String, Any> = emptyMap()
    ) : ToolResult() {
        override val isError = false
    }
    
    /**
     * 错误结果  
     */
    data class Error(
        val error: String,
        val code: Int = -1,
        override val content: List<ContentItem> = emptyList()
    ) : ToolResult() {
        override val isError = true
    }
    
    companion object {
        /**
         * 创建成功结果（文本内容）
         */
        fun success(text: String, metadata: Map<String, Any> = emptyMap()) = Success(
            content = listOf(ContentItem.text(text)),
            metadata = metadata
        )
        
        /**
         * 创建成功结果（结构化数据）
         */
        fun success(data: Any, metadata: Map<String, Any> = emptyMap()) = Success(
            content = listOf(ContentItem.json(data)),
            metadata = metadata
        )
        
        /**
         * 创建错误结果
         */
        fun error(message: String, code: Int = -1) = Error(
            error = message,
            code = code,
            content = listOf(ContentItem.text("错误: $message"))
        )
    }
}

/**
 * 内容项
 */
sealed class ContentItem {
    /**
     * 文本内容
     */
    data class Text(val text: String) : ContentItem()
    
    /**
     * JSON数据内容
     */
    data class Json(val data: JsonElement) : ContentItem()
    
    /**
     * 二进制内容
     */
    data class Binary(val data: ByteArray, val mimeType: String) : ContentItem()
    
    companion object {
        fun text(content: String) = Text(content)
        fun json(data: Any) = Json(JsonPrimitive(data.toString()))
        fun binary(data: ByteArray, mimeType: String) = Binary(data, mimeType)
    }
}

/**
 * 内部工具处理器
 */
internal data class ToolHandler(
    val name: String,
    val description: String,
    val parameterSchema: Map<String, ParameterInfo>? = null,
    val handler: suspend (Map<String, Any>) -> Any
) {
    fun toDefinition(): ToolDefinition {
        return if (parameterSchema != null) {
            ToolDefinition.withParameterInfo(name, description, parameterSchema)
        } else {
            ToolDefinition.simple(name, description)
        }
    }
}

File: vscode-extension/src/sdk/claude/mcp/mcpServer.ts
/**
 * MCP Server Interface - Base interface for all MCP servers
 * 
 * Translated from Kotlin: claude-agent-sdk/src/main/kotlin/com/asakii/claude/agent/sdk/mcp/McpServer.kt
 * 
 * Provides standard interface for creating custom MCP tool servers,
 * supporting tool listing and tool execution.
 */

/**
 * Parameter types for tool definitions
 */
export enum ParameterType {
    STRING = 'string',
    NUMBER = 'number',
    BOOLEAN = 'boolean',
    ARRAY = 'array',
    OBJECT = 'object'
}

/**
 * Parameter information - contains type and description
 */
export interface ParameterInfo {
    type: ParameterType;
    description?: string;
}

/**
 * JSON Schema for tool input
 */
export interface JsonSchema {
    type: string;
    properties?: Record<string, JsonSchema>;
    items?: JsonSchema;
    required?: string[];
    description?: string;
    enum?: (string | number | boolean)[];
    default?: unknown;
    [key: string]: unknown;
}

/**
 * Tool definition
 */
export interface ToolDefinition {
    name: string;
    description: string;
    inputSchema: JsonSchema;
}

/**
 * Tool definition helper functions
 */
export const ToolDefinitionHelper = {
    /**
     * Create simple tool definition with no parameters
     */
    simple(name: string, description: string): ToolDefinition {
        return {
            name,
            description,
            inputSchema: {
                type: 'object',
                properties: {}
            }
        };
    },

    /**
     * Create tool definition with parameter types (legacy compatibility)
     */
    withParameters(
        name: string,
        description: string,
        parameters: Record<string, ParameterType>
    ): ToolDefinition {
        const parameterInfo: Record<string, ParameterInfo> = {};
        for (const [key, type] of Object.entries(parameters)) {
            parameterInfo[key] = { type };
        }
        return this.withParameterInfo(name, description, parameterInfo);
    },

    /**
     * Create tool definition with parameter info (supports descriptions)
     */
    withParameterInfo(
        name: string,
        description: string,
        parameters: Record<string, ParameterInfo>
    ): ToolDefinition {
        const properties: Record<string, JsonSchema> = {};
        const required: string[] = [];

        for (const [paramName, paramInfo] of Object.entries(parameters)) {
            const schema: JsonSchema = {
                type: paramInfo.type
            };
            if (paramInfo.description) {
                schema.description = paramInfo.description;
            }
            properties[paramName] = schema;
            required.push(paramName);
        }

        return {
            name,
            description,
            inputSchema: {
                type: 'object',
                properties,
                required
            }
        };
    }
};

/**
 * Content item types
 */
export type ContentItem =
    | { type: 'text'; text: string }
    | { type: 'json'; data: unknown }
    | { type: 'binary'; data: Uint8Array; mimeType: string };

/**
 * Content item helper functions
 */
export const ContentItemHelper = {
    text(content: string): ContentItem {
        return { type: 'text', text: content };
    },
    json(data: unknown): ContentItem {
        return { type: 'json', data };
    },
    binary(data: Uint8Array, mimeType: string): ContentItem {
        return { type: 'binary', data, mimeType };
    }
};

/**
 * Tool execution result
 */
export type ToolResult = ToolResultSuccess | ToolResultError;

/**
 * Success result
 */
export interface ToolResultSuccess {
    isError: false;
    content: ContentItem[];
    metadata?: Record<string, unknown>;
}

/**
 * Error result
 */
export interface ToolResultError {
    isError: true;
    error: string;
    code?: number;
    content: ContentItem[];
}

/**
 * Tool result helper functions
 */
export const ToolResultHelper = {
    /**
     * Create success result with text content
     */
    success(text: string, metadata?: Record<string, unknown>): ToolResult {
        return {
            isError: false,
            content: [ContentItemHelper.text(text)],
            metadata
        };
    },

    /**
     * Create success result with JSON data
     */
    successJson(data: unknown, metadata?: Record<string, unknown>): ToolResult {
        return {
            isError: false,
            content: [ContentItemHelper.json(data)],
            metadata
        };
    },

    /**
     * Create success result with content items
     */
    successWithContent(content: ContentItem[], metadata?: Record<string, unknown>): ToolResult {
        return {
            isError: false,
            content,
            metadata
        };
    },

    /**
     * Create error result
     */
    error(message: string, code: number = -1): ToolResult {
        return {
            isError: true,
            error: message,
            code,
            content: [ContentItemHelper.text(`Error: ${message}`)]
        };
    }
};

/**
 * Tool handler function type
 */
export type ToolHandler = (arguments_: Record<string, unknown>) => Promise<ToolResult>;

/**
 * Internal tool handler interface
 */
export interface ToolHandlerBase {
    name: string;
    description: string;
    handler: ToolHandler;
    toDefinition(): ToolDefinition;
}

/**
 * Internal tool handler with parameter info
 */
export class ToolHandlerWithParams implements ToolHandlerBase {
    constructor(
        public readonly name: string,
        public readonly description: string,
        public readonly parameterSchema: Record<string, ParameterInfo> | null,
        public readonly handler: ToolHandler
    ) {}

    toDefinition(): ToolDefinition {
        if (this.parameterSchema) {
            return ToolDefinitionHelper.withParameterInfo(
                this.name,
                this.description,
                this.parameterSchema
            );
        }
        return ToolDefinitionHelper.simple(this.name, this.description);
    }
}

/**
 * Internal tool handler with full JSON schema
 */
export class ToolHandlerWithSchema implements ToolHandlerBase {
    constructor(
        public readonly name: string,
        public readonly description: string,
        public readonly inputSchema: JsonSchema,
        public readonly handler: ToolHandler
    ) {}

    toDefinition(): ToolDefinition {
        return {
            name: this.name,
            description: this.description,
            inputSchema: this.inputSchema
        };
    }
}

/**
 * MCP Server Interface - Base interface for all MCP servers
 */
export interface McpServer {
    /**
     * Server name, used as identifier
     */
    readonly name: string;

    /**
     * Server version
     */
    readonly version: string;

    /**
     * Server description
     */
    readonly description: string;

    /**
     * Tool call timeout in milliseconds
     * - null/undefined, 0, or negative: infinite timeout (for tools requiring user interaction)
     * - positive: specified timeout
     */
    readonly timeout?: number | null;

    /**
     * Whether to reset timeout timer when receiving progress reports
     * 
     * When true, timeout timer resets as long as MCP tool keeps reporting progress.
     * This allows long-running operations (like user interaction) to not timeout.
     * 
     * Reference: https://github.com/anthropics/claude-code/issues/470
     */
    readonly resetTimeoutOnProgress?: boolean;

    /**
     * Whether progress reporting is enabled
     * 
     * When true, MCP server should periodically report progress during long operations.
     * Used with resetTimeoutOnProgress to prevent timeouts.
     */
    readonly progressReportingEnabled?: boolean;

    /**
     * Get system prompt appendix for this MCP server
     * 
     * MCP servers can provide additional system prompts through this method,
     * which will be appended to the main system prompt to guide AI on how to
     * correctly use the tools provided by this server.
     * 
     * @returns System prompt appendix content, null means no additional prompt
     */
    getSystemPromptAppendix(): string | null;

    /**
     * Get list of tools that should be auto-approved
     * 
     * Tool names returned will be added to Claude CLI's --allowed-tools parameter,
     * these tools will automatically get permission without user confirmation.
     * 
     * Note: Return short tool names (e.g., "AskUserQuestion"),
     * system will automatically add mcp__{serverName}__ prefix.
     * 
     * @returns List of auto-approved tool names, default empty list
     */
    getAllowedTools(): string[];

    /**
     * List all available tools
     */
    listTools(): Promise<ToolDefinition[]>;

    /**
     * Call specified tool (Map arguments version, for legacy compatibility)
     */
    callTool(toolName: string, arguments_: Record<string, unknown>): Promise<ToolResult>;

    /**
     * Call specified tool with toolUseId
     * 
     * toolUseId is passed through async context, tools can get it via currentToolUseId().
     * Mainly used for file operation tools to record file history for showing Diff.
     * 
     * @param toolName Tool name
     * @param arguments_ Tool arguments
     * @param toolUseId Tool use ID (optional)
     * @returns Tool execution result
     */
    callToolWithContext(
        toolName: string,
        arguments_: Record<string, unknown>,
        toolUseId?: string | null
    ): Promise<ToolResult>;
}
